### PR TITLE
simplify the isColocateJoin function in PlanFragmentBuilder

### LINKED ISSUE #5037: cte join  lead to Must be replicate join or colocate join error
### Steps to reproduce the behavior (Required)
```
DROP DATABASE IF EXISTS n5;
CREATE DATABASE n5;
USE n5;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DATE NOT NULL ,c_0_1 BOOLEAN NOT NULL ,c_0_2 DATETIME NOT NULL ,c_0_3 VARCHAR(21) NOT NULL ,c_0_4 DECIMAL(19, 12)  NOT NULL ,c_0_5 INT NOT NULL ,c_0_6 CHAR(1) NULL ,c_0_7 DATETIME NULL ,c_0_8 SMALLINT NOT NULL ,c_0_9 DOUBLE NULL ,c_0_10 ARRAY<INT> NULL ,c_0_11 ARRAY<INT> NOT NULL ,c_0_12 ARRAY<STRING> NULL ,c_0_13 VARCHAR(1) NULL ,c_0_14 SMALLINT NOT NULL ,c_0_15 VARCHAR(1) NOT NULL , INDEX index1_c_0_2(`c_0_2`) USING BITMAP, INDEX index1_c_0_3(`c_0_3`) USING BITMAP) UNIQUE KEY (c_0_0,c_0_1,c_0_2,c_0_3) PARTITION BY RANGE(c_0_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 101 day)) DISTRIBUTED BY HASH (c_0_0,c_0_2) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t1 (c_1_0 DATETIME NOT NULL ,c_1_1 DATETIME NULL ,c_1_2 LARGEINT NOT NULL ,c_1_3 BOOLEAN NOT NULL ,c_1_4 CHAR(21) NULL ,c_1_5 BIGINT NOT NULL ,c_1_6 LARGEINT NOT NULL ,c_1_7 BOOLEAN NULL ,c_1_8 FLOAT NULL ,c_1_9 CHAR(11) NULL ) UNIQUE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4) PARTITION BY RANGE(c_1_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 1095 day)) DISTRIBUTED BY HASH (c_1_1,c_1_3,c_1_0,c_1_2,c_1_4) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t2 (c_2_0 DATETIME NOT NULL ,c_2_1 DECIMAL(17, 13)  NOT NULL ,c_2_2 BOOLEAN REPLACE NULL ) AGGREGATE KEY (c_2_0,c_2_1) PARTITION BY RANGE(c_2_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 136 day)) DISTRIBUTED BY HASH (c_2_0) BUCKETS 3 properties("replication_num"="3") ;
CREATE TABLE t3 (c_3_0 DATETIME NULL ) UNIQUE KEY (c_3_0) PARTITION BY RANGE(c_3_0) (START ("2010-01-01") END ("2021-12-31") EVERY (INTERVAL 175 day)) DISTRIBUTED BY HASH (c_3_0) BUCKETS 3 properties("replication_num"="3") ;
DELETE FROM t0 WHERE c_0_3 IS NULL AND c_0_2="1970-01-22 23:31:09" AND c_0_1=true;
INSERT INTO t0 VALUES ("2010-01-14", false, "2010-01-31 08:00:14", "", 0.000002863152, 30, "", "2010-01-29 08:00:14", 5, 0.54413047114609, [-489647138, 1508575369, -1338895239, 1298012187, 1418468302, -1611457693, -990004405, -1364899140, 636961214, 1332930684], [-1906890962, 1025628303, 1162260838, -1061101406, -1710690429, 224188180], [], "", 3, ""), ("2010-01-22", true, "2010-01-29 08:00:14", "", 0.000006261857, 31, "", "2010-01-26 08:00:14", 12, 0.36739912178478573, [-692871042], [208214618, 1, -240887220, -763335365, 1753546324, -268118343, -2072771509], ["QpgO9rNzc", "a", "P", "tEa", "Wivdhq", "oWnNRlrZl", ""], NULL, 8, "");
INSERT INTO t1 VALUES ("2010-01-10 08:00:14", NULL, 1999965732, false, NULL, -1587661401, -569687717, false, 0.10138960440394906, "KZu"), ("2010-01-16 08:00:14", NULL, 229801448, false, "", -2086700885, -2072365226, NULL, NULL, ""), ("2010-01-26 08:00:14", "2010-01-27 08:00:14", -1727259651, true, "Iy", 72214622, 2065787896, true, 0.4907633327624811, "");
INSERT INTO t1 VALUES ("2010-01-30 08:00:14", NULL, -922326926, false, "GSTHrEGGJ", 1251454987, -839512849, false, 0.8395907660206663, "JtD"), ("2010-01-03 08:00:14", "2010-01-22 08:00:14", -1893223035, false, "co7g", -811637495, -804470759, true, 0.20010876688764556, ""), ("2010-01-19 08:00:14", "2010-01-24 08:00:14", -1737318714, false, "", 1585386663, 864923574, true, 0.46192801419610385, "z");
INSERT INTO t0 VALUES ("2010-01-10", false, "2010-01-17 08:00:14", "jQkgTY2", 0.000007493295, 16, "", "2010-01-09 08:00:14", 11, NULL, [], [1687029214, -1321805809, -1030824751, 1421473388, 156238196, -1714116582, 926527669, -708207862], ["F", "D", "zOd", "", "Z", ""], "", 13, "");
INSERT INTO t2 VALUES ("2010-01-14 08:00:14", 4.971E-10, false), ("2010-01-17 08:00:14", 8.847E-10, false);
INSERT INTO t0 VALUES ("2010-01-09", true, "2010-01-18 08:00:14", "K", 3.2664E-8, 16, "", "2010-01-19 08:00:14", 15, 0.5666710379051416, [1844878156, 324560111], [-1497918359, -1647694741], ["sk", "F", "aEKgAc", "c", "l", "i", "sB2H66y5", "Ek", "W", "", "6", "", ""], "", 2, "");
INSERT INTO t2 VALUES ("2010-01-13 08:00:14", 7.675E-10, true), ("2010-01-06 08:00:14", 1.073E-10, NULL), ("2010-01-29 08:00:14", 8.623E-10, false);
INSERT INTO t1 VALUES ("2010-01-22 08:00:14", "2010-01-19 08:00:14", -409440293, true, NULL, -1413891908, -1080313057, true, NULL, NULL), ("2010-01-29 08:00:14", "2010-01-30 08:00:14", -1050032514, false, NULL, -290341078, 307106169, false, 0.15719879326254205, NULL), ("2010-01-21 08:00:14", "2010-01-03 08:00:14", 1976272040, false, NULL, -1081754178, 283261254, false, -2.77865529E8, "AYETe");
INSERT INTO t0 VALUES ("2010-01-01", true, "2010-01-26 08:00:14", "", 0.000008132189, 20, "", "2010-01-11 08:00:14", 8, -2.139533959E9, [-1174099174, -554468059, -1369659494, -2016483485, 1304417178, -601628810, -1994555446, 121957788, 1141898983, -1857389280, -1407437751, 28447045], [1881118140, 2138305711, 1525849833, -701722913, -610846984], ["I", "hAg0", "VWwgb", "HsIYlXIC", "7", "r5MXXqf", "RM", "U", "XYq", "m", "rmk"], "", 12, "");
INSERT INTO t0 VALUES ("2010-01-30", false, "2010-01-01 08:00:14", "", 0.000006692845, 9, "", "2010-01-13 08:00:14", 0, 0.10277318170166949, [1310198099, 835486402, -768078490, -1136082420, 1321375765, -1589681897, -190412839], [1999965732, 1523071250, 1157812225, 91791619], ["McyqJY955"], "", 8, ""), ("2010-01-01", false, "2010-01-24 08:00:14", "", 0.000007349419, 1, "", "2010-01-22 08:00:14", 1, 0.20010876688764556, [-587048798, -811637495, -2021761873, 1694894896, 1517085671, -863844162, -271913680, -2029677505, -1178388053, -9223372036854775808, 753531705, -919969581, 270139336, -877341578], [-1190141489, 1673917628, -7042012, -773785056, 728717411], [], "", 14, ""), ("2010-01-01", false, "2010-01-31 08:00:14", "HPj", 1.0697E-8, 17, "", "2010-01-05 08:00:14", 4, -6.10846984E8, [1220827242, -502265417, -1133596499, -153478253, 2111668145, -496798394, -51803204, -123464927, 1521768710, -1173580074, 1728617886, 1398560019], [-499991465, -817277083, 728717411, 998045774, -1267570016, -2030727655, 872205654, 141797127, 1018524318], ["", "C4nHExu", "", "", "0wCtv7", "i", ""], "", 6, "");
INSERT INTO t2 VALUES ("2010-01-17 08:00:14", 5.928E-10, true), ("2010-01-04 08:00:14", 4.678E-10, false);
INSERT INTO t3 VALUES ("2010-01-04 08:00:14"), ("2010-01-01 08:00:14");
INSERT INTO t1 VALUES ("2009-12-31 08:00:14", "2010-01-12 08:00:14", 1471198112, false, "", -601628810, -352892619, false, 0.9475529005975245, "u"), ("2010-01-11 08:00:14", "2010-01-08 08:00:14", -247814338, false, "CyfrX6pD", 1679405355, -1057140378, false, 0.6181853073886651, NULL);
INSERT INTO t1 VALUES ("2010-01-01 08:00:14", "2009-12-31 08:00:14", 1162848525, false, "4x", 1356185851, 9223372036854775807, false, NULL, NULL), ("2010-01-05 08:00:14", "2010-01-29 08:00:14", 1321375765, false, "3fMy", -151908605, 455057655, true, 0.35944530182367607, NULL), ("2010-01-28 08:00:14", "2010-01-07 08:00:14", -770434896, false, "", 158491963, -372088128, false, NULL, "");
INSERT INTO t3 VALUES ("2010-01-18 08:00:14"), ("2010-01-05 08:00:14");
INSERT INTO t1 VALUES ("2010-01-15 08:00:14", "2010-01-21 08:00:14", -1854888898, false, "nn", -832056987, 1593262438, false, -1.53478253E8, "");


mysql> WITH with_t_0 as (SELECT t3.c_3_0 FROM t3 WHERE ((t3.c_3_0) IS NOT NULL)) SELECT COUNT(*) FROM  (SELECT t3.c_3_0 FROM t3) subt3 FULL OUTER JOIN with_t_0 ON subt3.c_3_0 = with_t_0.c_3_0 AND subt3.c_3_0 = CAST(with_t_0.c_3_0 AS DATE )  AND subt3.c_3_0 != with_t_0.c_3_0 AND with_t_0.c_3_0 != CAST(subt3.c_3_0 AS DATE )  WHERE ((subt3.c_3_0) IS NULL);
ERROR 1064 (HY000): Must be replicate join or colocate join

```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - branch-2.2 Release
 - mysql> select current_version();
+---------------------------------+
| current_version()               |
+---------------------------------+
| SQLANCER_01_RELEASE 145a79c |
+---------------------------------+

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5037 

## Problem Summary(Required) ：

The origin is isColocateJoin is complex and error-prone, use the required property to decide is more simple

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 9 / 10 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|9|10|90.00%|

**Review by Seaven — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Review by liuyehcf — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 1]
```diff
@@ -1835,66 +1834,21 @@ public PlanFragment visitPhysicalHashJoin(OptExpression optExpr, ExecPlan contex
...
+            // through the required properties type check if it is colocate join
+            return optExpression.getRequiredProperties().stream().allMatch(
+                    physicalPropertySet -> {
+                        if (!physicalPropertySet.getDistributionProperty().isShuffle()) {
+                            return false;
+                        }
+                        HashDistributionDesc.SourceType hashSourceType =
+                                ((HashDistributionSpec) (physicalPropertySet.getDistributionProperty().getSpec()))
+                                        .getHashDistributionDesc().getSourceType();
+                        if (hashSourceType.equals(HashDistributionDesc.SourceType.LOCAL)) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1847]: why not return in one-line?

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 9 / 10 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|9|10|90.00%|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by Youngwb** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**